### PR TITLE
DM-14574 Implement SimpleMagnitudeModel and related magnitude infrastructure

### PR DESCRIPTION
**Pull request by parejkoj**

### CODE REVIEW

File: python/lsst/jointcal/jointcal.py [hunk 1]
```diff
@@ -596,14 +597,17 @@ def _fit_photometry(self, associations, dataName=None):
...
             model = lsst.jointcal.ConstrainedPhotometryModel(associations.getCcdImageList(),
                                                              self.focalPlaneBBox,
                                                              visitOrder=self.config.photometryVisitOrder)
             # potentially nonlinear problem, so we may need a line search to converge.
             doLineSearch = self.config.allowLineSearch
-        elif self.config.photometryModel == "simple":
-            model = lsst.jointcal.SimplePhotometryModel(associations.getCcdImageList())
+        elif self.config.photometryModel == "simpleFlux":
+            model = lsst.jointcal.SimpleFluxModel(associations.getCcdImageList())
+            doLineSearch = False  # purely linear problem, so no line search needed
```
Comment (TallJimbo) [inline python/lsst/jointcal/jointcal.py]: I don't think the simple flux model is linear - you have FittedStar parameters and observational parameters that get multiplied together, so it's linear in any *single* parameter, but for a problem to be linear it needs to be linear in the parameter *vector*.

File: src/SimplePhotometryModel.cc [hunk 2]
```diff
@@ -1,4 +1,5 @@
 #include <iostream>
+#include <math.h>
```
Comment (TallJimbo) [inline src/SimplePhotometryModel.cc]: c++ code should always use `#include <cmath>` instead.  That may mean you need to add some `std::` in front of math functions, but you should do that anyway.

File: src/SimplePhotometryModel.cc [hunk 3]
```diff
@@ -106,5 +80,73 @@ PhotometryMappingBase *SimplePhotometryModel::findMapping(CcdImage const &ccdIma
...
+double SimpleFluxModel::transformError(CcdImage const &ccdImage, MeasuredStar const &star) const {
+    auto mapping = findMapping(ccdImage);
+    return mapping->transformError(star, star.getInstFluxErr());
+}
+
+std::shared_ptr<afw::image::PhotoCalib> SimpleFluxModel::toPhotoCalib(CcdImage const &ccdImage) const {
+    double calibration = (findMapping(ccdImage)->getParameters()[0]);
+    auto oldPhotoCalib = ccdImage.getPhotoCalib();
+    return std::unique_ptr<afw::image::PhotoCalib>(
+            new afw::image::PhotoCalib(calibration, oldPhotoCalib->getCalibrationErr()));
```
Comment (TallJimbo) [inline src/SimplePhotometryModel.cc]: So jointcal isn't updating the estimated uncertainty in the calibration?

Comment (TallJimbo) [reply src/SimplePhotometryModel.cc]: Also, we're in C++14 now, so you can use `std::make_unique()`.

File: include/lsst/jointcal/BaseStar.h [hunk 4]
```diff
@@ -64,16 +64,22 @@ class BaseStar : public FatPoint {
     double getFluxErr() const { return _fluxErr; }
     void setFluxErr(double fluxErr) { _fluxErr = fluxErr; }
 
+    double getMag() const { return _mag; }
+    double &getMag() { return _mag; }
+    void setMag(double mag) { _mag = mag; }
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply include/lsst/jointcal/BaseStar.h]: How would you suggest handling this sort of situation, then? I recall asking about this before, and the reference getter was the generally preferred solution, I though.

```
fittedStar->getFlux() += mi->getFlux();
```

File: src/FittedStar.cc [hunk 5]
```diff
@@ -39,8 +31,8 @@ void FittedStar::setRefStar(const RefStar *refStar) {
 }
 
 void FittedStar::addMagMeasurement(double magValue, double magWeight) {
-    _mag = (_mag * _wmag + magValue * magWeight) / (_wmag + magWeight);
-    _wmag += magWeight;
+    _mag = (_mag * _magErr + magValue * magWeight) / (_magErr + magWeight);
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply src/FittedStar.cc]: Sadly, clang-format puts the spaces around every operator, and I have clang-format configured to operate on save.

@pschella : is there a way to get clang-format to be more particular about its operator spacing rules?

File: src/SimplePhotometryModel.cc (hunk 3 above)
Comment (parejkoj) [reply src/SimplePhotometryModel.cc]: Correct: I haven't worked through the math about what the new PhotoCalib uncertainty is. The input uncertainty for that CCD Exposure's Calib object is close enough for the moment, I think?

File: include/lsst/jointcal/BaseStar.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (TallJimbo) [reply include/lsst/jointcal/BaseStar.h]: Given that you actually use the reference getter with e.g. that `+=` line, I think I'd remove the setter, but if you think there's a chance you'd ever use a different internal representation for this class (i.e. only hold flux and compute magnitude on the fly), you should only keep the setter and rewrite `+=` lines accordingly.

File: src/SimplePhotometryModel.cc (hunk 3 above)
Comment (TallJimbo) [reply src/SimplePhotometryModel.cc]: Yeah, this is not a priority for photometry until we get the constrained photometry fit working.  And I suppose we don't actually have a place to put the astrometric uncertainty, though maybe we should think about how to propagate that as well.

File: include/lsst/jointcal/BaseStar.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (parejkoj) [reply include/lsst/jointcal/BaseStar.h]: Sometimes I need to compute the magnitude separately and set it, depending on the calibration that is used for the given source. So, I have the fitter doing e.g. `fittedStar.getMag() -= delta;` when doing offsets, and the initialization doing using the setter (the star itself doesn't have an associated photoCalib):

```auto mag = _photoCalib->instFluxToMagnitude(ms->getInstFlux(), ms->getInstFluxErr(), point);
ms->setMag(mag.value);
```

File: include/lsst/jointcal/BaseStar.h [hunk 6]
```diff
@@ -27,10 +27,10 @@ class BaseStar : public FatPoint {
...
     };
     //! constructor
-    BaseStar(double xx, double yy, double flux, double fluxErr)
-            : FatPoint(xx, yy), _flux(flux), _fluxErr(fluxErr){};
-    BaseStar(Point const &point, double flux, double fluxErr)
-            : FatPoint(point), _flux(flux), _fluxErr(fluxErr){};
+    BaseStar(double xx, double yy, double flux, double fluxErr, double mag, double magErr)
+            : FatPoint(xx, yy), _flux(flux), _fluxErr(fluxErr), _mag(mag), _magErr(magErr){};
+    BaseStar(Point const &point, double flux, double fluxErr, double mag, double magErr)
+            : FatPoint(point), _flux(flux), _fluxErr(fluxErr), _mag(mag), _magErr(magErr){};
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply include/lsst/jointcal/BaseStar.h]: After playing around with it for a bit, it looks like you're right. I thought I needed to do those calculations differently in different places, but it's already managed correctly everywhere. I've removed those, and move the `magFromFlux` and `magErrFromFluxErr` into `BaseStar.h`.

File: src/Gtransfo.cc [hunk 7]
```diff
@@ -439,10 +439,10 @@ GtransfoPoly::GtransfoPoly(const Gtransfo *gtransfo, const Frame &frame, unsigne
     double step = sqrt(fabs(frame.getArea()) / double(nPoint));
     for (double x = frame.xMin + step / 2; x <= frame.xMax; x += step)
         for (double y = frame.yMin + step / 2; y <= frame.yMax; y += step) {
-            auto pix = std::make_shared<BaseStar>(x, y, 0, 0);
+            auto pix = std::make_shared<BaseStar>(x, y, 0, 0, 0, 0);
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply src/Gtransfo.cc]: Fixed as part of the above mag/magErr removal.

File: python/lsst/jointcal/jointcal.py (hunk 1 above)
Comment (parejkoj) [reply python/lsst/jointcal/jointcal.py]: Good point. I'll clarify it to say that the model parameters are purely linear, which I'm pretty sure is the cause of our trouble and need for the lineSearch.

File: src/SimplePhotometryModel.cc (hunk 3 above)
Comment (parejkoj) [reply src/SimplePhotometryModel.cc]: Agreed on all points.

**Pull request merged**